src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -310,6 +310,110 @@ public void DoAbort()
 
             public abstract void InvokeCallback(bool allowPooling);
 
+            internal virtual bool TryBatch(SocketAsyncContext context, int id, ref Interop.Sys.IoControlBlock ioControlBlock) => false;
+
+            internal virtual void HandleBatchEvent(in Interop.Sys.IoEvent ioControlBlock)
+            {
+                Debug.Fail("Expected to be implemented only by types overriding TryAsBatch");
+                throw new InvalidOperationException();
+            }
+
+            protected bool CanRetry(SocketError error) => error == SocketError.TryAgain || error == SocketError.WouldBlock;
+
+            // todo: move this method somewhere else
+            protected static SocketError Map(Interop.Error error)
+            {
+                switch (error)
+                {
+                    case Interop.Error.EACCES:
+                        return SocketError.AccessDenied;
+                    case Interop.Error.EADDRINUSE:
+                        return SocketError.AddressAlreadyInUse;
+                    case Interop.Error.EADDRNOTAVAIL:
+                        return SocketError.AddressNotAvailable;
+                    case Interop.Error.EAFNOSUPPORT:
+                        return SocketError.AddressFamilyNotSupported;
+                    case (Interop.Error)11:
```

Comment:
important: this method requires a rewrite - `AIO` is Linux system call and it returns different error codes that what we have in `Interop.Error`. A good example is `Interop.Error.EAGAIN` which is not 11 (while here it must be 11)

Reply:
It would make sense for this to be in the pal `IoGetEvents`. The errnos returned can be mapped to `IoEvent` in the `IoEvent`.

Reply:
@tmds very good idea!

Reply:
@adamsitnik @tmds my understanding is that none of these errors codes are mapped correctly.
`io_event.Res` will contain [Linux error codes](https://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html) which have nothing to do with the [`pal_errno.h` codes starting with `0x1`](https://github.com/dotnet/runtime/blob/master/src/libraries/Native/Unix/System.Native/pal_errno.h#L33).  I'm planning to map them in `IoGetEvents` as suggested, reusing the [common error mapping logic](https://github.com/dotnet/runtime/blob/master/src/libraries/Native/Unix/System.Native/pal_errno.c#L20).
Please correct me if my understanding is wrong!

Note: currently I'm debugging a case of `ESPIPE` (29 - "Illegal seek").

Reply:
> Please correct me if my understanding is wrong!

@antonfirsov You are correct